src/main.rs

```diff
@@ -205,41 +273,93 @@ async fn get_new_bech32_address(client: &mut tonic_lnd::Client) -> Address {
         .expect("lnd returned invalid address")
 }
 
-#[tokio::main]
-async fn main() -> Result<(), Box<dyn std::error::Error>> {
-    let (config, mut args) = Config::including_optional_config_files(std::iter::empty::<&str>()).unwrap_or_exit();
+fn process_args<A: Iterator<Item = OsString>>(
+    args: A,
+) -> Result<Option<ScheduledPayJoin>, ArgError> {
+    // ensure all args are utf8
+    let args = args
+        .map(|arg| arg.into_string())
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(ArgError::NotUTF8)?;
+
+    // first argument is fee rate
+    let fee_rate = match args.get(0) {
+        Some(fee_rate_str) => fee_rate_str
+            .parse::<u64>()
+            .map_err(ArgError::FeeRateError)?,
+        None => return Ok(None),
+    };
 
-    let client = tonic_lnd::connect(config.lnd_address, &config.lnd_cert_path, &config.lnd_macaroon_path)
-        .await
-        .expect("failed to connect");
+    // the remaining argument is the wallet amount (if any)
+    let mut wallet_amount = None;
+
+    // parse scheduled channel arguments: pairs of (addr, amount)
+    let channels = match args.get(1..) {
+        Some(args) => args
+            .chunks(2)
+            .filter_map(|chunk| match chunk.len() {
+                // parse scheduled channel arg: a pair of (addr, amount)
+                2 => Some(ScheduledChannel::from_args(&chunk[0], &chunk[1])), // p2p_address, btc_amount
+                // parse wallet amount (optional)
+                1 => {
+                    wallet_amount = Some(
+                        bitcoin::Amount::from_str_in(
+                            chunk[0].as_str(),
+                            bitcoin::Denomination::Satoshi,
+                        )
+                        .map_err(ArgError::InvalidWalletAmount),
+                    );
+                    None
+                }
+                unexpected_len => panic!("chunk len should only be 1 or 2, got {}", unexpected_len),
+            })
+            .collect::<Result<VecDeque<_>, _>>()?,
+        None => return Ok(None),
+    };
```

Comment:
I like the comments that explain what is going on. However I think we should keep the, the original loop + match with `args.next()`. There's way less nesting and fewer lines. Can we merge the best of both and keep the comments but use the original logic?

Reply:
@DanGould I tried again, this is my new solution. Let me know what you think.

```rust
    // parse scheduled channel arguments: pairs of (addr, amount)
    let mut channels = VecDeque::with_capacity(args.len() / 2);

    // the remaining single argument is the wallet amount in satoshis (if any)
    let wallet_amount = loop {
        match (args.next(), args.next()) {
            // we have a pair of arguments, interpret it as a scheduled channel (p2p addr, amount)
            (Some(addr), Some(amount)) => {
                channels.push_back(ScheduledChannel::from_args(addr, amount)?);
            }
            // if there is a remaining single argument, it is the wallet amount
            (Some(amount), None) => {
                break bitcoin::Amount::from_str_in(amount, bitcoin::Denomination::Satoshi)
                    .map_err(ArgError::InvalidWalletAmount)?
            }
            // if there is no remaining single argument, the wallet amount is 0
            _ => break bitcoin::Amount::ZERO,
        }
    };
```

Reply:
beautiful!